### DISCUSSION TITLE
Function not implemented

### DISCUSSION THREAD

**Discussion by f4r4** [cesanta/mongoose#1786] [Ask questions, get answers] [answered]
Could any one point to a documentation article that describes about this error 
0 3 mongoose.c:4209:read_conn           2 0x4 0:0 288 err 38 (Function not implemented)

its a simple rest server and handling the connection with in this if condition on a fn function
```
else if (ev == MG_EV_HTTP_MSG) 

mongoose.c:3565:mg_listen           1 0x3 http://0.0.0.0:8010
0 3 mongoose.c:4316:accept_conn         2 accepted 192.168.10.40:33424
0 3 mongoose.c:4209:read_conn           2 0x4 0:0 288 err 38 (Function not implemented)
0 3 mongoose.c:4220:write_conn          2 0x4 251:0 251 err 38 (Function not implemented)
0 3 mongoose.c:4209:read_conn           2 0x4 0:0 -1 err 38 (Function not implemented)
0 3 mongoose.c:3520:mg_close_conn       2 closed


```

please advice

**Comment by scaprile** [comment]
Can you please tell us on which one of the supported architectures, with what configuration, you are trying to run your piece of code, using which Mongoose version, with what macro options enabled ?

**Reply by f4r4** [reply]
1. PowerPC
2. No configs 
3. No macro versions enabled 
4. v7.7

I understand its not in the supported architectures, if you could point me where this error happens 
or definitions of the error code I can try to debug things 

```

#include <stdio.h>

#include "mongoose.h"
#include "mjson/mjson.h"

static const char *s_listen_url = "http://0.0.0.0:8010";
static void fn(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
  if (ev == MG_EV_HTTP_MSG) {
    struct mg_http_message *hm = (struct mg_http_message *) ev_data;
    if (mg_http_match_uri(hm, "/")) {
      char *s = NULL;
      mjson_printf(&mjson_print_dynamic_buf, &s, "{\"result\": \"%s\"}\n", "2.0.0");
      mg_http_reply(c, 200, "Content-Type: application/json\r\n", s);
      free(s);       
    }
  } else if (ev == MG_EV_CLOSE) {
    c->is_closing = 1;
  }
  (void) fn_data;
}

int main(void) {
  struct mg_mgr mgr;

  mg_log_set("3");                       // Set log level
  mg_mgr_init(&mgr);                             // Initialise event manager
  mg_http_listen(&mgr, s_listen_url, fn, NULL);  // Start proxy
  for (;;) mg_mgr_poll(&mgr, 1000);              // Event loop
  mg_mgr_free(&mgr);

  return 0;
}

```

**Reply by scaprile** [reply]
https://github.com/cesanta/mongoose/blob/d5993ba27ece4b406c230eca63b76dd5d2c28a2d/mongoose.c#L4205-L4209
https://github.com/cesanta/mongoose/blob/d5993ba27ece4b406c230eca63b76dd5d2c28a2d/mongoose.c#L4215-L4220
`mg_sock_recv()` and `mg_sock_send()`
https://github.com/cesanta/mongoose/blob/d5993ba27ece4b406c230eca63b76dd5d2c28a2d/mongoose.c#L4072-L4077
your `send()` socket function is returning that error code.
Which OS are you using ? If you don't tell Mongoose the architecture, it will try to guess it
https://github.com/cesanta/mongoose/blob/7.7/src/arch.h
and complain if it can't.
It seems to be guessing some architecture you seem to don't have fully functional as the socket interface is not working properly.
This seems to be the documentation you need: https://mongoose.ws/documentation/#build-options

**Comment by scaprile** [comment, marked as answer]
This seems to be the documentation you need: https://mongoose.ws/documentation/#build-options

**Comment by cpq** [comment]
@f4r4 

1. Use latest version - and you don't need mjson.c and mjson.h. Mongoose has all JSON functionality built-in
2. On susccessful reads, and your are successful, just ignore the error message
